### PR TITLE
Add a editable SkinEditor Element List

### PR DESCRIPTION
**Pull request by C0D3-M4513R**
This pr will attempt to implement #21098 .

done:
 - [x] Add the initial list implementation
 - [x] Sync the list contents to the drawables onscreen
 - [x] Add Lists that can be minimised
 - [x] Allow items in the List to be selected from the editor.
 - [x] Allow items to be selected from the List.
 - [x] Fix naming of the elements in the editor list
 - [x] Add a drag and drop handler, or another way to move IDrawableListItem's
 - [x] Link the changes to elements onscreen
 - [x] Allow those changes to be saved and loaded to config
 - [x] ~Fix having to set a fixed height for the OsuScrollContainer inside DrawableList/RearrangeableListContainer~ I do not think, that this is possible to fix. Please correct me, if I'm wrong.
 - [x] Optional: Rework DrawableListItem to simplify Selection Logic
 - [x] Optional: Allow Nested lists. ~This would also require adding a way to group together items in the editor, since I need a representative item for the DrawableMinimizableList, if I wanted to add it to a DrawableList (due to how RearrangeableListContainer is structured).~ 
 - [x] When Switching Screens the names don't get instantly updated
 - [x] Add Tests
 - [x] Rework how Lists are minimised. I do not like the OsuCheckbox for this. Arrows might work, when looking at how the enum selector works.
 - [x] Test drag on nested lists.
 - [x] Fix the Static Height in DrawableList
 - [x] Raise some sort of event (which includes old, and new parent) when changing parents of elements
 - [x] Should `OnDragAction` continue to be customisable? - No, it shouldn't.

optional: 
 - Rework, how to highlight selected items in a list - still looks completly fine, and may be easier to read currently.
 - Add user creatable nested lists? - potentially not this pr

### CODE REVIEW

File: osu.Game/Screens/Edit/List/DrawableContainer.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Shapes;
+using osu.Game.Graphics.UserInterface;
+using osuTK;
+
+namespace osu.Game.Screens.Edit.List
+{
+    public class DrawableContainer<T> : CompositeDrawable, IDrawableListItem<T>
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/DrawableContainer.cs:15]: This class name needs consideration.... It just can't be this.

File: osu.Game/Screens/Edit/List/DrawableList.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input.Events;
+using osu.Game.Graphics.UserInterface;
+using osuTK;
+using osuTK.Input;
+
+namespace osu.Game.Screens.Edit.List
+{
+    public class DrawableList<T> : CompositeDrawable, IDrawableListItem<T>
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/DrawableList.cs:15]: Is this something like `RearrangeableListContainer`? You might want to check whether you can already lean on an existing implementation like that.

**Comment by peppy** [comment]
I think the general direction is good here, but it's very hard to review this change as it stands.

Here's a few pointers:
- How are you testing this? I'd hope to see some kind of test scene coverage to show how this interacts when there's more than one layer (and potentially nested layers).
- You mentioned having to give scroll container a height: this is correct. You probably want to make it `RelativeSizeAxes = Axes.Y` to have it expand to the containing grid container.
- I get that the checkbox to expand the list is temporary, but it's going to have to be replaced to get this in because it looks super weird at the moment. Maybe just use a simple `SpriteIcon` `+` and `-` (from fontawesome) for now?

File: osu.Game/Screens/Edit/List/DrawableList.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            //todo: compute this somehow add runtime
+            Height = 100f;
+            ListContainer.Spacing = new Vector2(2.5f);
+            SelectAll = ((IDrawableListItem<T>)this).SelectableOnStateChanged;
+        }
+
+        protected override bool OnClick(ClickEvent e)
+        {
+            if (e.Button == MouseButton.Left)
+                SelectAll(SelectionState.NotSelected);
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/DrawableList.cs:72]: This reads super weird. "select all in a not selected state"?

File: osu.Game/Screens/Edit/List/DrawableList.cs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        {
+            if (drawables is null) return;
+
+            foreach (T drawable in drawables)
+            {
+                addInternal(drawable);
+            }
+
+            OnDragAction();
+        }
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/DrawableList.cs:87]: All this stuff seems too custom. I'd expect the list to just be a `Container` and handle things that way.

File: osu.Game/Screens/Edit/List/DrawableList.cs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+        public bool Select(T drawable, bool value = true)
+        {
+            if (!ItemMap.ContainsKey(drawable)) return false;
+
+            return select(ItemMap[drawable], value);
+        }
+
+        public virtual void Select(bool value)
+        {
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/DrawableList.cs:156]: These overrides make no sense. One takes a drawable and one just..  takes a bool?

File: osu.Game/Screens/Edit/List/IDrawableListItem.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                    SelectInternal(false);
+                    return;
+            }
+        }
+
+        public LocalisableString GetDefaultText(Drawable target)
+        {
+            Logger.Log("GetDefaultText with" + target + " target.");
+            return target.Name.Equals(string.Empty) ? (target.GetType().DeclaringType ?? target.GetType()).Name : target.Name;
+        }
```
Comment (peppy) [inline osu.Game/Screens/Edit/List/IDrawableListItem.cs:47]: We generally avoid interface level default methods.

**Review by peppy — CHANGES_REQUESTED** [review]
Very brief pass over code quality

**Comment by peppy** [comment]
The general code quality here isn't up to scratch with the rest of the project. I think it may pay to read around some of the other classes we have and get a bit more used to the code base. I think this is going to be hard to move forward at this rate.

**Comment by peppy** [comment]
This is starting to feel quite usable.

Immediate note is that when you click a top level container and it selects all content, the item's visible selection state doesn't update after deselecting one (or all) of the elements:


https://user-images.githubusercontent.com/191335/204985875-c41b6c8d-0be7-4812-bcd3-61636b545953.mp4

**Comment by Feodor0090** [comment]
Something is wrong here.

Add something (default acc used) and something else (black box used)
изображение

Black box is above everything, and it's in the bottom of list.

Then try to reorder it:
изображение
изображение

Now item in the bottom of list is under everything.

I checked "default" elements, they are ordered as "highest in list is the top in container":
изображение

Not sure which element should be above everything - top one or bottom one. Ignoring this, the only todo here is to add new elements in list on proper position (with current ordering to head, not to the end).

Blueprint ordering issue seems to be fixed by last commit.

**Comment by Feodor0090** [comment]
Now new things are under everything. Aside from subjective point that this is weird, it changes behaviour (before 2 last commits new ones were on top). You should probably fix this.
изображение

**Comment by Feodor0090** [comment]
Just noticed more weird thing: notice how elements are flipping their order each gameplay enter.

https://user-images.githubusercontent.com/53872073/206876770-2c2aab1c-41dd-46b9-a87f-9a9a8f89cdc7.mp4

**Comment by Feodor0090** [comment]
Could not find any more ways to break this yet.

Two more nitpicks:
1) margins:
изображение
2) add something and a black box. Try to drag black box in the list. Notice that slider and checkbox flickers each mouse movement. Seems that the settings panel is recreated each drag. Would be nice not to do it.

**Comment by peppy** [comment]
This PR is unfortunately in a very hard place. The code style and sheer size of the PR (lines of code wise) makes it very hard to even start reviewing it. On the flip side, this *is* a component that we need for the skin editor, to both support depth adjustments and also in the future to support moving components between layers.

I don't really have any more to say for now, just thought I'd mention this.

**Pull request closed without merging**